### PR TITLE
Add retry to failed document text reads

### PR DESCRIPTION
**Pull request by mattwar**
Potential/partial fix for #1360 

The retry logic lets file reads succeed after hitting temporary read failures, or temporary missing files. If still failing after nominal number of retries, then continue to use existing behavior of logging a failure and returning empty text.

@Pilchie @jasonmalinowski please review

note: some tests were changed because they did not have the files they were trying to open, or they were opening files when they did not need too.

### CODE REVIEW

**Comment by Pilchie** [comment]
LGTM.

**Comment by mattwar** [comment]
@dotnet-bot please test this

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs [hunk 1]
```diff
@@ -107,6 +114,12 @@ protected static ValueSource<TextAndVersion> CreateRecoverableText(TextLoader lo
             }
             catch (IOException e)
             {
+                if (++retries <= MaxRetries)
```
Comment (pharring) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs]: FileNotFoundException is-a IOException. Should it be treated differently (no retries)?I don't really want to wait 5 seconds for that.

Comment (mattwar) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs]: @pharring it is a common scenario that tools such as designers rewriting documents may indeed cause them to appear deleted before rewritten.  In such a case, file watchers can trigger a reload of the document while it is missing.

Also, the max wait is only 1 sec.

**Comment by pharring** [comment]
:+1:

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs [hunk 2]
```diff
@@ -90,8 +90,15 @@ protected static ValueSource<TextAndVersion> CreateRecoverableText(TextLoader lo
                 services.TemporaryStorage);
         }
 
+        private const int MaximumRetryDelay = 1000; // 1 second
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs]: Nitpick: state the unit,  or even better if possible just make this a TimeSpan.

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs [hunk 3]
```diff
@@ -107,6 +114,12 @@ protected static ValueSource<TextAndVersion> CreateRecoverableText(TextLoader lo
             }
             catch (IOException e)
             {
+                if (++retries <= MaxRetries)
+                {
+                    await Task.Delay(RetryDelay).ConfigureAwait(false);
+                    goto retry;
+                }
+
                 services.Workspace.OnWorkspaceFailed(new DocumentDiagnostic(WorkspaceDiagnosticKind.Failure, e.Message, documentId));
```
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs]: Should there be some sort of logging that retries are happening?

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs (hunk 1 above)
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs]: What about UnauthorizedAccessException? Or are we already wrapping all non-IOExceptions with IOException deep down in the text loader?

**Comment by jasonmalinowski** [comment]
:+1:

**Comment by mattwar** [comment]
@dotnet-bot test this pleast

**Comment by mattwar** [comment]
@dotnet-bot test this please

**Comment by jaredpar** [comment]
@dotnet-bot  test this please

**Pull request merged**